lib/math/enum.ex

```diff
@@ -82,24 +83,55 @@ defmodule Math.Enum do
 
   def median(collection) do
     count = Enum.count(collection)
+    mid_point = div(count, 2)
 
     cond do
       count == 0 ->
         nil
 
       # Middle element exists
-      rem(count, 2) == 1 ->
-        Enum.sort(collection) |> Enum.at(div(count, 2))
+      Integer.is_odd(count) ->
+        collection
+        |> Enum.sort()
+        |> Enum.fetch!(mid_point)
 
       true ->
-        # Take two middle-most elements.
-        sorted_collection = Enum.sort(collection)
-
-        [
-          Enum.at(sorted_collection, div(count, 2)),
-          Enum.at(sorted_collection, div(count, 2) - 1)
-        ]
+        collection
+        |> Enum.sort()
+        |> Enum.slice((mid_point - 1)..mid_point)
         |> Math.Enum.mean()
     end
   end
+
+  @doc """
+  Calculates the mode of a given collection of numbers.
+
+  Always returns a list. An empty input results in an empty list. Supports bimodal/multimodal collections by returning a list with multiple values.
+
+  ## Examples
+
+      iex> Math.Enum.mode [1, 2, 3, 4, 1]
+      [1]
+      iex> Math.Enum.mode [1, 2, 3, 2, 3]
+      [2, 3]
+      iex> Math.Enum.mode []
+      []
+      # iex> Math.Enum.mode [1, 1.0, 2, 3]
+      # [1]
+      # iex> Math.Enum.mode [1, 1.0, 2, 3] &===/1
+      # [1, 1.0]
+  """
+  @spec mode(Enum.t()) :: Enum.t()
+  def mode(collection)
+
+  def mode(collection) do
+    collection
+    |> Enum.reduce(%{}, fn k, acc -> Map.update(acc, k, 0, &(&1 + 1)) end)
+    |> Enum.group_by(&elem(&1, 1), &elem(&1, 0))
+    |> Enum.max_by(&elem(&1, 0), fn -> [] end)
+    |> (fn
```

Comment:
Looks very nice!

I would however propose to replace the inline anonymous function with a private function that is called.

Or, even better:

``` 
    |> Enum.max_by(&elem(&1, 0), fn -> {{}, []} end)
    |> elem(1)
```
which also skips the pattern-match :grinning: !

Reply:
:man_facepalming: